### PR TITLE
Changes to gem5 Resources [gem5 Vision]

### PR DESCRIPTION
**Pull request by kunpai**
We added a config file to configs/
We changes resources in src/python/gem5/resources by adding an api folder with mongoclient and jsonclient
We also made corresponding changes to testing directory and resource python files to make it work with our new schema

### CODE REVIEW

File: configs/gem5Resources.config.json
Comment (powerjg) [inline configs/gem5Resources.config.json:1]: Shouldn't go in configs. A gem5 "config" is different than this file.

File: src/python/gem5/components/boards/kernel_disk_workload.py [hunk 1]
```diff
@@ -94,30 +94,30 @@ def get_disk_device(self) -> str:
         raise NotImplementedError
 
     @abstractmethod
-    def _add_disk_to_board(self, disk_image: DiskImageResource) -> None:
+    def _add_disk_to_board(self, diskimage: DiskImageResource) -> None:
```
Comment (powerjg) [inline src/python/gem5/components/boards/kernel_disk_workload.py]: Why did you change this variable name? Please revert unless absolutely necessary.

File: src/python/gem5/resources/api/client.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        :param resources: A list of resources to search through.
+        :param resource_version: The version of the resource to search for.
+        :return: The resource object as a Python dictionary if found.
+        :raises: RuntimeError if the resource with the specified version is not found.
+        """
+        for resource in resources:
+            if resource["resource_version"] == resource_version:
+                if core.gem5Version not in resource["gem5_versions"]:
+                    warnings.warn(
+                        f"Resource compatible with gem5 version: '{core.gem5Version}' not found.\n"
```
Comment (powerjg) [inline src/python/gem5/resources/api/client.py]: Line length.

File: src/python/gem5/resources/api/client.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        """
+        Searches for the resource with the given version. If the resource is
+        not found, an exception is thrown.
+        :param resources: A list of resources to search through.
+        :param resource_version: The version of the resource to search for.
+        :return: The resource object as a Python dictionary if found.
+        :raises: RuntimeError if the resource with the specified version is not found.
+        """
+        for resource in resources:
+            if resource["resource_version"] == resource_version:
```
Comment (powerjg) [inline src/python/gem5/resources/api/client.py]:
```suggestion
resource = next([resource for resource in resources if resource["resource_version"] == resource_version), None)
if not resource:
    raise ...
if core.gem5Version not in resource["gem5_versions"]:
    ...
```

I don't love the nestedness.

File: src/python/gem5/resources/api/client.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        Returns a list of compatible resources with the given gem5 version.
+        :param resources: A list of resources to filter.
+        :return: A list of compatible resources as Python dictionaries.
+        If no compatible resources are found, the original list of resources
+        is returned.
+        """
+        compatible_resources = []
+        for resource in resources:
+            if core.gem5Version in resource["gem5_versions"]:
+                compatible_resources.append(resource)
```
Comment (powerjg) [inline src/python/gem5/resources/api/client.py]:
```suggestion
        compatible_resources = [resource for resource in resources if core.gem5Version in resource["gem5_versions"]]
```

File: src/python/gem5/resources/api/client.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+        if len(compatible_resources) == 0:
+            warnings.warn(
+                f"Resource compatible with gem5 version: '{core.gem5Version}' not found.\n"
+                "Resource versions can be found at: "
+                f"https://gem5vision.github.io/gem5-resources-website/resources/{resources[0]['id']}/versions"
+            )
+            return resources
```
Comment (powerjg) [inline src/python/gem5/resources/api/client.py]: Add a comment about why you are returning *all* the resources here.

File: src/python/gem5/resources/api/client_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        else:
+            clients[resource] = JSONClient(database["url"])
+    return clients
+
+
+# read gem5/configs/gem5Resources.config.json using pathlib
+config = json.load(
+    open(
+        Path(__file__).parent.parent.parent.parent.parent.parent.parent
+        / "configs/gem5Resources.config.json",
```
Comment (powerjg) [inline src/python/gem5/resources/api/client_wrapper.py]: Do you mean for this filename to be hardcoded? If so, can you add more comments? Otherwise, it should be a variable (maybe an environment variable?)

Also, I don't think that we should use this file's location for hardcoding this config. @BobbyRBruce and I can think about the better way to do this.

File: src/python/gem5/resources/api/client_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+from .jsonclient import JSONClient
+from .mongoclient import MongoClient
+import json
+from pathlib import Path
+from typing import Optional, Dict, Union, Type, Tuple, List
+
+
+def create_clients(
+    config: dict,
+):
```
Comment (powerjg) [inline src/python/gem5/resources/api/client_wrapper.py]: Needs some documentation

File: src/python/gem5/resources/api/client_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+# read gem5/configs/gem5Resources.config.json using pathlib
+config = json.load(
+    open(
+        Path(__file__).parent.parent.parent.parent.parent.parent.parent
+        / "configs/gem5Resources.config.json",
+        "r",
+    )
+)
+
+clients = create_clients(config)
```
Comment (powerjg) [inline src/python/gem5/resources/api/client_wrapper.py]: I'm a bit uncomfortable with this code running on import.

File: configs/example/gem5_library/riscvmatched-hello.py
Comment (powerjg) [inline configs/example/gem5_library/riscvmatched-hello.py:1]: This should be a separate change.

File: src/python/gem5/resources/api/jsonclient.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        elif not self._url_validator(self.path):
+            raise Exception(
+                f"Resources location '{self.path}' is not a valid path or URL."
+            )
+        else:
+            self.resources = requests.get(self.path).json()
+
+    def get_resource_obj(
+        self, resource_id: str, resource_version: Optional[str] = None
+    ) -> Dict[str, Union[str, Dict[str, str], List[str]]]:
```
Comment (powerjg) [inline src/python/gem5/resources/api/jsonclient.py]: I'm not sure the return type is correct here, but it's not a big deal. Maybe simply `Dict[str, Any]`

File: src/python/gem5/resources/api/jsonclient.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        :param resource_id: The ID of the Resource.
+        :optional param resource_version: The version of the Resource.
+        If not given, the latest version compatible with current gem5 version is returned.
+        :return: The Resource as a Python dictionary.
+        """
+        # getting all the resources with the given id from MongoDB
```
Comment (powerjg) [inline src/python/gem5/resources/api/jsonclient.py]: Typo with MongoDB here.

File: src/python/gem5/resources/api/mongoclient.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from .client import AbstractClient
+import requests
```
Comment (powerjg) [inline src/python/gem5/resources/api/mongoclient.py]: As far as I can tell, this is *not* in python's standard library. See https://requests.readthedocs.io/en/latest/user/install/#install

That said... you can probably replace it with something from the stdlib. https://docs.python.org/3/library/urllib.request.html#module-urllib.request

File: src/python/gem5/resources/api/mongoclient.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        :return: The Resource as a Python dictionary.
+        """
+        # getting all the resources with the given ID from MongoDB
+        resources = list(self.__get_resource_by_id(resource_id))
+        # if no resource with the given ID is found throw an Exception
+        if len(resources) == 0:
+            raise Exception(f"Resource with ID '{resource_id}' not found.")
+        # sorting the resources by version
+        resources.sort(
+            key=lambda x: list(map(int, x["resource_version"].split("."))),
```
Comment (powerjg) [inline src/python/gem5/resources/api/mongoclient.py]: You could also use this to sort by semantic versions. I don't have a strong preference, though.

https://stackoverflow.com/questions/2574080/sorting-a-list-of-dot-separated-numbers-like-software-versions

File: src/python/gem5/resources/api/client_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def get_resource_obj(
+    resource_id,
+    resource_version: Optional[str] = None,
+    database: Optional[str] = None,
```
Comment (powerjg) [inline src/python/gem5/resources/api/client_wrapper.py]: I'm confused by this function's name. It's the same as `AbstractClient`, but it has a different signature. 

I suggest not having the same function names with different signatures.

File: src/python/gem5/resources/downloader.py [hunk 14]
```diff
@@ -523,7 +524,6 @@ def get_resource(
             with tarfile.open(download_dest) as f:
 
                 def is_within_directory(directory, target):
-
```
Comment (powerjg) [inline src/python/gem5/resources/downloader.py:526]: Don't arbitrarily remove whitespace. (I'm surprised that black wasn't run)

File: src/python/gem5/resources/looppoint.py [hunk 15]
```diff
@@ -491,7 +491,6 @@ def __init__(
         with open(_path) as file:
             json_contents = json.load(file)
             for rid in json_contents:
-
```
Comment (powerjg) [inline src/python/gem5/resources/looppoint.py:494]: No whitespace changes.

File: tests/pyunit/stdlib/resources/pyunit_md5_utils_check.py [hunk 16]
```diff
@@ -75,7 +75,6 @@ class MD5DirTestSuite(unittest.TestCase):
     """Test cases for gem5.resources.md5_utils.md5_dir()"""
 
     def _create_temp_directory(self) -> Path:
-
```
Comment (powerjg) [inline tests/pyunit/stdlib/resources/pyunit_md5_utils_check.py:78]: whitespace....

File: src/python/gem5/resources/resource.py [hunk 17]
```diff
@@ -27,6 +27,8 @@
 from abc import ABCMeta
 import os
 from pathlib import Path
+
+# from m5.util import warn, fatal
```
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: Don't comment things out. Delete them or keep them.

File: src/python/gem5/resources/resource.py [hunk 18]
```diff
@@ -522,9 +555,11 @@ def _get_weights_and_simpoints_from_file(
 
 
 def obtain_resource(
-    resource_name: str,
+    resource_id: str,
     resource_directory: Optional[str] = None,
     download_md5_mismatch: bool = True,
+    resource_version: Optional[str] = None,
+    database: Optional[str] = None,
```
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: Needs to be documented in the docstring.

File: src/python/gem5/resources/resource.py [hunk 19]
```diff
@@ -547,12 +582,13 @@ def obtain_resource(
     """
 
     # Obtain the JSON resource entry for this resource
-    resource_json = get_resources_json_obj(resource_name)
+    resource_json = get_resource_obj(
```
Comment (powerjg) [inline src/python/gem5/resources/resource.py]: Should this now be something like `resource_data` since it's just a dictionary?

File: .gitignore
Comment (powerjg) [inline .gitignore:1]: Should be in a separate change.

**Review by powerjg — COMMENTED** [review]
Overall, this looks really good. @BobbyRBruce, I tagged you in one place. Otherwise, I think it would be best for you to spend your time looking at the testing code. I didn't look at the tests at all.

There's one "high-level" thing I don't like, though I guess I could be convinced it's ok. I don't like that you changed `disk_image` to `diskimage`. This seems like an arbitrary change that just adds to the diff and doesn't actually improve anything. I would prefer to keep the name `disk_image`.

Also, look through the diff and remove any whitespace-only changes.

Finally, what is the backwards compatibility story? Do we need to have a DB running or an entirely new json file ready before merging?

**Review by powerjg — CHANGES_REQUESTED** [review]
Overall, this looks really good. @BobbyRBruce, I tagged you in one place. Otherwise, I think it would be best for you to spend your time looking at the testing code. I didn't look at the tests at all.

There's one "high-level" thing I don't like, though I guess I could be convinced it's ok. I don't like that you changed `disk_image` to `diskimage`. This seems like an arbitrary change that just adds to the diff and doesn't actually improve anything. I would prefer to keep the name `disk_image`.

Also, look through the diff and remove any whitespace-only changes.

Finally, what is the backwards compatibility story? Do we need to have a DB running or an entirely new json file ready before merging?

File: configs/gem5Resources.config.json
Comment (kunpai) [reply configs/gem5Resources.config.json:1]: Where is a good place we can place it?

Comment (powerjg) [reply configs/gem5Resources.config.json:1]: The root would be a better location. However, Bobby and I need to chat about how to resolve that location from the stdlib. I don't like the `__file_path__.parent.parent.parent...` thing.

File: .gitignore [hunk 20]
```diff
@@ -32,3 +32,5 @@ configs/example/memcheck.cfg
 configs/dram/lowp_sweep.cfg
 .pyenv
 .vscode
+
```
Comment (BobbyRBruce) [inline .gitignore]: Is there a good reason to have an empty line here?

File: configs/gem5Resources.config.json [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+            "dataSource": "gem5-vision",
+            "database": "gem5-vision",
+            "collection": "versions_test",
+            "url": "https://data.mongodb-api.com/app/data-ejhjf/endpoint/data/v1",
+            "name": "data-ejhjf",
+            "apiKey": "OIi5bAP7xxIGK782t8ZoiD2BkBGEzMdX3upChf9zdCxHSnMoiTnjI22Yw5kOSgy9",
+            "isMongo": true
+        }
+    }
+}
```
Comment (BobbyRBruce) [inline configs/gem5Resources.config.json]: This is petty in the extreme but it's pretty standard to have a newline character at the end of a file. See here for why: https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

File: src/python/gem5/resources/api/client.py [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+            raise Exception(
+                f"Resource {resources[0]['id']} with version '{resource_version}'"
+                " not found.\nResource versions can be found at: "
+                f"https://gem5vision.github.io/gem5-resources-website/resources/{resources[0]['id']}/versions"
+            )
+        if core.gem5Version not in resource["gem5_versions"]:
+            warnings.warn(
+                "Resource compatible with gem5 version: "
+                f"'{core.gem5Version}' not found.\n"
+                "Resource versions can be found at: "
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: Expand this warning a little bit. As is written it doesn't state what this means for your simulation run.

I'd say: "Resource {resource['id']} version {resource_version} is not known to be compatible with gem5 version {core.gem5Version}. This may cause problems with your simulation. This resource's compatibility with different gem5 versions can be found here: ..."

This makes it clear that this might be dangerous but the simulator isn't going to stop you.

File: src/python/gem5/resources/api/client.py [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+                "https://gem5vision.github.io/gem5-resources-website"
+                f"/resources/{resources[0]['id']}/versions"
+            )
+        return resource
+
+    def _get_compatible_resources(self, resources: List) -> List:
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: I'd change this function to return an empty list if no compatible resources are found in the list. I know why you're doing this but I feel like this function should be simple and correspond to its name.

For example, if all the resources in the list are compatible with the current version of gem5, then this function's return is identical to the case where none of them are. We're losing information.

File: src/python/gem5/resources/api/client.py [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        Returns a list of compatible resources with the given gem5 version.
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: We're phrasing here "given gem5 version". I'd rephrase to "current gem5 version".

File: src/python/gem5/resources/api/client.py [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+        :param url: The URL to be validated.
+        :return: True if the URL is valid, False otherwise.
+        """
+        try:
+            result = urllib.parse.urlparse(url)
+            return all([result.scheme, result.netloc, result.path])
+        except:
+            return False
+
+    def _search_version(self, resources: List, resource_version: str) -> dict:
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: This function should be merged into `_get_compatible_resources`. Perhaps this function can be:

```
def _get_resources_comatible(resources: : List[dict], resource_version: str =core.gem5Version ) -> List[dict]:
```

It will filter the list to those only with a compatible version.

File: src/python/gem5/resources/api/client.py [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+                    resource
+                    for resource in resources
+                    if resource["resource_version"] == resource_version
+                ]
+            ),
+            None,
+        )
+        if resource is None:
+            raise Exception(
+                f"Resource {resources[0]['id']} with version '{resource_version}'"
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: I think these warnings/exceptions should be put elsewhere. In this function, as it it written, we cannot assume all the resources passed have the same id. This function could, theoretically, be used to filter all the resources we have to the list of those with known compatibility (this may be useful at some point).

I would probably opt for putting these in a separate function with a signature like:

```
check_resource_version_capatibility(resource: dict, gem5_version: str = core.gem5Version) ->None
```

You could call it when you want to run these checks.

File: src/python/gem5/resources/api/client.py [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+            # which contains all the versions of the resource
+            warnings.warn(
+                f"Resource compatible with gem5 version: '{core.gem5Version}' not found.\n"
+                "Resource versions can be found at: "
+                f"https://gem5vision.github.io/gem5-resources-website/resources/{resources[0]['id']}/versions"
+            )
+            return resources
+        return compatible_resources
+
+    def _sort_resources_by_version(self, resources: List) -> List:
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client.py]: In keeping with my other comments, this signature here does not restrict a user from inputting a list of resources with different IDs. I'd change this function to be `_sort_resources`, which sorts resources based first on id, then on version.

File: src/python/gem5/resources/api/client_wrapper.py [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+def create_clients(
+    config: dict,
+):
+    """
+    This function creates respective client object for each database in the config file according to the type of database.
+    Params: config: config file containing the database information
+    Returns: clients: dictionary of clients for each database
+    """
+    clients = {}
+    for resource in config["resources"]:
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client_wrapper.py]: Is it too late to call call this something like "clients" instead of "resources". Needless to say, the word "resources" here is getting overused.

File: src/python/gem5/resources/api/client_wrapper.py [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+from .jsonclient import JSONClient
+from .mongoclient import MongoClient
+import json
+from pathlib import Path
+from typing import Optional, Dict, Union, Type, Tuple, List
+
+
+def create_clients(
+    config: dict,
+):
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client_wrapper.py]: Add return type here 'Dict'

File: src/python/gem5/resources/api/client_wrapper.py [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    This function returns the resource object from the corresponding database.
+
+    :param resource_id: resource id of the resource
+    :optional param resource_version: resource version of the resource
+    :optional param database: database name. If not provided, the first database in the config file is used
+    :return: resource object
+    """
+    if not database:
+        database = list(clients.keys())[0]
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client_wrapper.py]: I'm not sure if we have discussed this before, but is this the behavior we want? If the user doesn't specify the database here, we just assume we're looking in the first one.

How about this: we don't have a "database" field, and we always search all our clients for instances of a resource. We throw an exception if there are duplicates (same id and version). Would that work?

File: src/python/gem5/resources/api/jsonclient.py [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+from pathlib import Path
+import requests
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/jsonclient.py]: As Jason said in the other file, I don't think we can use requests.

File: src/python/gem5/resources/api/jsonclient.py [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+        if Path(self.path).is_file():
+            self.resources = json.load(open(self.path))
+        elif not self._url_validator(self.path):
+            raise Exception(
+                f"Resources location '{self.path}' is not a valid path or URL."
+            )
+        else:
+            self.resources = requests.get(self.path).json()
+
+    def get_resource_obj_from_client(
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/jsonclient.py]: It seems to me like a simpler way of doing this would be to have an abstract function in `AbstractClient` called `get_resources_by_id(id: str)`. Then have `get_resource_obj_from_client` implemented in `AbstractClient` (which calls `get_resources_by_id`).

This would mean `JSONClient` and `MongoClient` would only need to implement `get_resources_by_id` which seems to be the only thing that truly differs between the two clients.

File: src/python/gem5/resources/downloader.py [hunk 33]
```diff
@@ -109,7 +110,6 @@ def _get_resources_json_at_path(path: str, use_caching: bool = True) -> Dict:
     # its Truncated Exponential Backoff algorithm
     # (maximum of roughly 1 minute). Typically this code will run quickly.
     with FileLock(f"{download_path}.lock", timeout=120):
-
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/downloader.py:112]: You've deleted a line here for no real reason. Avoid doing stuff like this. It just makes the change a bit messier than it needs to be.

File: src/python/gem5/resources/downloader.py [hunk 34]
```diff
@@ -41,6 +41,7 @@
 from urllib.error import HTTPError
 from typing import List, Dict, Set, Optional
 
+from python.gem5.resources.api.client_wrapper import get_resource_obj
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/downloader.py]: Within the stdlib you should use relative imports. So in this case, `from .api.client_wrapper import get_resource_obj`

File: src/python/gem5/resources/looppoint.py [hunk 35]
```diff
@@ -491,7 +491,7 @@ def __init__(
...
+
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/looppoint.py:494]: This to me shows you probably don't have our pre-commit checks installed. It will automatically detect and remove unnecessary whitespace like this. Run `./util/pre-commit-install.sh` to do so

File: src/python/gem5/resources/workload.py [hunk 36]
```diff
@@ -217,5 +225,4 @@ def __init__(
             for key in workload_json["additional_params"]:
                 assert isinstance(key, str)
                 params[key] = workload_json["additional_params"][key]
-
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/workload.py]: I'll only say this once more, because there's numerous parts this happens, but please don't remove lines randomly like this. The commit/pull-request should only be focused on adding/removing code as needed.

File: src/python/SConscript [hunk 37]
```diff
@@ -268,6 +268,10 @@ PySource('gem5.resources', 'gem5/resources/resource.py')
 PySource('gem5.resources', 'gem5/resources/workload.py')
 PySource('gem5.resources', 'gem5/resources/looppoint.py')
 PySource('gem5.resources', 'gem5/resources/elfie.py')
+PySource('gem5.resources.api', 'gem5/resources/api/client_wrapper.py')
```
Comment (BobbyRBruce) [inline src/python/SConscript]: I think "api" is too generic. Can we call this directory "client_api" instead?

File: src/python/gem5/resources/downloader.py [hunk 38]
```diff
@@ -430,17 +432,17 @@ def get_resource(
     # minutes.Most resources should be downloaded and decompressed in this
     # timeframe, even on the most constrained of systems.
     with FileLock(f"{to_path}.lock", timeout=900):
-
-        resource_json = get_resources_json_obj(resource_name)
+        resource_obj = get_resource_obj(
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/downloader.py]: I'm getting a bit confused by this: is this returning a resource object (i.e. a full implementation such as `BinaryResource`, or a dictionary representing the JSON?). It's really not clear.

From what I can see in the rest of this file, it's just a JSON representation of the object, so why the change of this function name and variable. This seems pointless.

File: src/python/gem5/resources/api/client_wrapper.py [hunk 39]
```diff
@@ -0,0 +1,87 @@
...
+clients = None
+
+
+if clients is None:
+    # read gem5/configs/gem5Resources.config.json using pathlib
+    # gem5Resources.config.json contains the database information
+    config = json.load(
+        open(
+            Path(__file__).parent.parent.parent.parent.parent.parent.parent
+            / "configs/gem5Resources.config.json",
```
Comment (BobbyRBruce) [inline src/python/gem5/resources/api/client_wrapper.py]: There is an issue here of where the config file should exist, how it should be used, and how a user interacts with it. I discussed this with Jason and I think you should do the following:

1. The config file should be called "gem5-config.json".
2. It should be moved to "src/python".
3. This configs file should be compiled into the binary (me and Jason's answer to the question "but how do I do this?" is "🤷‍♂️" , but we _think_ it should be possible...)
4. This compiled-in config file config file will serve as the default values for those fields. In addition to this the user can set an environment variable to specify their own "gem5-config.json" file to override this. (e.g. "GEM5_CONFIG="/home/bbruce/my-config.json" ./build/ALL/gem5.opt ...` should allow for loading of the "my-config.json" file, replacing the compiled in version).
5. If there is a "gem5-config.json" in the current working directory, gem5 should use this.

4 and 5 should be optional stretch-goals for now (i.e., only do them if you've got nothing better to do and want to). but I think you should try to do 1, 2, and 3.

**Review by BobbyRBruce — COMMENTED** [review]
I haven't reviewed the tests (i'll do so another time).

Even though this looks like a lot of criticism, this is really excellent work. Just a couple more iterations of this and I think we'll be ready to incorporate into the gem5 project.
